### PR TITLE
Isolated optimizer

### LINKED ISSUE #50: Create optimizer.py file
I think we should create an ```optimpizer.py``` file one class per optimizer, especially we could add a simple gradient descent for the resolution of the bilevel optimition problem, as well as an adam optimizer.
Wdyt?

### PR DESCRIPTION
**Pull request by QB3**
The goal of this PR is to decouple the outer optimization process from the ```grad_search``` function.
This paves the way for naive gradient descent implementation for debugging, as well more refined optimizer such as adam.

closes #50

### CODE REVIEW

File: sparse_ho/optimizers/adam.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import numpy as np
+
+
+class Adam():
+    """This Adam code is taken from XXX
+
+    n_outer: int, optional (default=100).
+        number of maximum updates of alpha.
+    epsilon= float, optional (default=1e-3)
```
Comment (Klopfe) [inline sparse_ho/optimizers/adam.py:9]: epsilon :

File: sparse_ho/optimizers/adam.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            t += 1
+            value_outer, grad = _get_val_grad(log_alpha, self.tol, monitor)
+
+            if self.verbose:
+                print("Value outer criterion: %f" % value_outer)
+
+            if (i > 1) and (monitor.objs[-1] > monitor.objs[-2]):
+                break
+            # print("%i / %i  || crosss entropy %f  || accuracy %f" % (
+            #       i, n_outer, val, monitor.acc_vals[-1]))
```
Comment (Klopfe) [inline sparse_ho/optimizers/adam.py:54]: Do we want to leave this print commented? maybe have a verbose option?

File: sparse_ho/optimizers/line_search_wolfe.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+from numpy.linalg import norm
+
+
+class LineSearchWolfe():
+
```
Comment (Klopfe) [inline sparse_ho/optimizers/line_search_wolfe.py:5]: Comments for documentation

File: sparse_ho/optimizers/adam.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import numpy as np
+
+
+class Adam():
```
Comment (agramfort) [inline sparse_ho/optimizers/adam.py:4]: it should inherit from base class

**Pull request merged**